keras_rcnn/preprocessing/_object_detection.py

```diff
@@ -1,118 +1,84 @@
 
     def next(self):
+        # Lock indexing to prevent race conditions
         with self.lock:
-            index_array, current_index, current_batch_size = next(
-                self.index_generator)
-
-        index = index_array[0]
+            selection, _, batch_size = next(self.index_generator)
 
-        pathname = self.dictionary[index]["filename"]
+        # Transformation of images is not under thread lock so it can be done in parallel
+        image_batch    = numpy.zeros((batch_size,) + self.target_shape, dtype=keras.backend.floatx())
+        gt_boxes_batch = numpy.zeros((batch_size, 0, 5),                dtype=keras.backend.floatx())
+        metadata       = numpy.zeros((batch_size, 3),                   dtype=keras.backend.floatx())
 
-        image = skimage.io.imread(pathname)
+        for batch_index, image_index in enumerate(selection):
+            path         = self.data[image_index]["filename"]
+            image        = skimage.io.imread(path, as_grey=(self.target_shape[2] == 1))
+            image, scale = scale_image(image)
+            image        = self.image_data_generator.random_transform(image)
+            image        = self.image_data_generator.standardize(image)
 
-        image = numpy.expand_dims(image, 0)
+            # Copy image to batch blob
+            image_batch[batch_index] = image
 
-        ds = self.dictionary[index]["boxes"]
+            # Set ground truth boxes
+            boxes = self.data[image_index]["boxes"]
+            for i, b in enumerate(boxes):
+                if b["class"] not in self.classes:
+                    raise Exception("Class {} not found in '{}'.".format(b["class"], self.classes))
 
-        boxes = numpy.asarray([[d[k] for k in ['y1', 'x1', 'y2', 'x2']] for d in ds])
+                gt_data = [b["y1"], b["x1"], b["y2"], b["x2"], self.classes[b["class"]]]
+                gt_boxes_batch = numpy.append(gt_boxes_batch, [[gt_data]], axis=1)
 
-        labels = numpy.asarray([d['class'] for d in ds])
+            # Scale the ground truth boxes to the selected image scale
+            gt_boxes_batch[batch_index, :, :4] *= scale
 
-        labels = self.encoder.transform(labels)
-
-        labels = keras.utils.to_categorical(labels)
-
-        metadata = list(image.shape[1:-1]) + [1]
-
-        # boxes = numpy.expand_dims(boxes, 0)
-
-        # labels = numpy.expand_dims(labels, 0)
-
-        # metadata = numpy.expand_dims(metadata, 0)
-
-        return [image, boxes, labels, metadata], [boxes, labels]
+            # Create metadata
+            metadata[batch_index, :] = [image.shape[0], image.shape[1], scale]
 
+        return [image_batch, gt_boxes_batch, metadata], None
 
 class ObjectDetectionGenerator:
-    def __init__(self, classes):
-        self.classes = classes
-
-    def flow(self, dictionary, shuffle=True, seed=None):
-        return DictionaryIterator(dictionary, self, shuffle=shuffle, seed=seed)
+    def flow(self, data, classes, image_shape):
```

Comment:
We should drop the `input_shape` parameter.

`shape` should be inferred for situations where images have variable dimensions.